diff --git a/node_modules/types-mediawiki/mw/Api.d.ts b/node_modules/types-mediawiki/mw/Api.d.ts
index ffbe4d0..f970036 100644
--- a/node_modules/types-mediawiki/mw/Api.d.ts
+++ b/node_modules/types-mediawiki/mw/Api.d.ts
@@ -16,7 +16,7 @@ type Tail<T extends any[]> = T extends [] ? T : T extends [any?, ...infer R] ? R
 type TypeOrArray<T> = T extends any ? T | T[] : never; // T[] would be a mixed array
 type ReplaceValue<T extends U | U[], U, V> = T extends U[] ? V[] : V;
 
-type UnknownApiParams = Record<string, string | number | boolean | string[] | number[] | undefined>;
+export type UnknownApiParams = Record<string, string | number | boolean | string[] | number[] | undefined>;
 
 export type ApiResponse = Record<string, any>; // it will always be a JSON object, the rest is uncertain ...
 
diff --git a/node_modules/types-mediawiki/mw/config.d.ts b/node_modules/types-mediawiki/mw/config.d.ts
index c28faf2..eab787d 100644
--- a/node_modules/types-mediawiki/mw/config.d.ts
+++ b/node_modules/types-mediawiki/mw/config.d.ts
@@ -2,6 +2,26 @@ import { ExtensibleMap } from "./Map";
 
 declare global {
     namespace mw {
+        export interface DiscussionToolsPageThread {
+            type: 'heading' | 'comment';
+            level: number;
+            id: string;
+            replies: Reply[];
+        }
+
+        export interface DiscussionToolsHeading extends DiscussionToolsPageThread {
+            type: 'heading';
+            headingLevel: number;
+            name: string;
+        }
+
+        export interface DiscussionToolsReply extends DiscussionToolsPageThread {
+            type: 'comment';
+            timestamp: string;
+            author: string;
+            displayName?: string;
+        }
+
         /**
          * Map of configuration values.
          *
@@ -360,6 +380,8 @@ declare global {
              * @see https://www.mediawiki.org/wiki/Special:MyLanguage/Manual:Interface/JavaScript#wgDiffNewId
              */
             wgDiffNewId?: number;
+
+            wgDiscussionToolsPageThreads?: DiscussionToolsPageThread[];
         }>;
     }
 }
